File: test/survivorship_policy_test.exs
# test/survivorship_policy_test.exs — gr-6y2: Survivorship is policy-driven (rankings always on).
#
# Proves the engine seam that keeps medipim-specific scoring OUT of the generic core:
#   * %Priority{}  -> tier ranking, unchanged (back-compat).
#   * injected fn  -> context-aware rank (medipim's off-product penalty lives HERE, not in Priority).
# Attribute rankings are ALWAYS applied — there is no "off" for attribute survivorship (the toggle
# we want lives on identity-conflict guarding instead; see test/identity_conflict_explained_test.exs).
defmodule SurvivorshipPolicyTest do
  use ExUnit.Case, async: true

  defp e(source, value, order), do: %{source: source, value: value, order: order}

  test "back-compat: a %Priority{} still ranks by tier" do
    priority = Priority.new(%{"name" => [["orgA"], ["orgB"]]}, [])
    d = Survivorship.decide("name", [e("orgA", "Foo", 1), e("orgB", "Bar", 2)], priority)
    assert d.winner == "orgA"
    assert d.value == "Foo"
    assert d.status == :resolved
  end

  test "injected rank fn expresses medipim's off-product penalty (context-aware)" do
    scores = %{"name" => %{"A" => 10, "B" => 5}}

    # rank fn closes over the product's scoring org-set; an off-product source is devalued to -1.
    rank = fn scoring_orgs ->
      fn dim, src ->
        base = get_in(scores, [dim, src]) || 0
        score = if MapSet.member?(scoring_orgs, src), do: base, else: -1
        -score
      end
    end

    entries = [e("B", "keep", 1), e("X", "drop", 2)]

    # Product P1: X is NOT on the product -> penalised below default-0/on-product sources -> B wins.
    p1 = rank.(MapSet.new(["A", "B"]))
    assert Survivorship.decide("name", entries, p1).winner == "B"

    # Product P2: X IS on the product -> scores default 0, B (off-product here) is penalised -> X wins.
    # Same claims, different context, different winner — exactly what one static Priority could not do.
    p2 = rank.(MapSet.new(["X"]))
    assert Survivorship.decide("name", entries, p2).winner == "X"
  end
end

File: php/tests/SurvivorshipPolicyTest.php
<?php

declare(strict_types=1);

namespace Ingot\Tests;

use Ingot\Priority;
use Ingot\Survivorship;
use PHPUnit\Framework\TestCase;

/**
 * PHP-parity mirror of test/survivorship_policy_test.exs (gr-6y2): Survivorship is policy-driven.
 * A Priority ranks by tier (back-compat); an injected callable carries context-aware scoring
 * (medipim's off-product penalty) without touching the core. Attribute rankings are always applied.
 */
final class SurvivorshipPolicyTest extends TestCase
{
    /** @param mixed $value */
    private static function e(?string $source, $value, int $order): array
    {
        return ['source' => $source, 'value' => $value, 'order' => $order];
    }

    public function test_backcompat_priority_ranks_by_tier(): void
    {
        $priority = Priority::new(['name' => [['orgA'], ['orgB']]], []);
        $d = Survivorship::decide('name', [self::e('orgA', 'Foo', 1), self::e('orgB', 'Bar', 2)], $priority);

        self::assertSame('orgA', $d['winner']);
        self::assertSame('Foo', $d['value']);
        self::assertSame('resolved', $d['status']);
    }

    public function test_injected_rank_fn_expresses_off_product_penalty(): void
    {
        $scores = ['name' => ['A' => 10, 'B' => 5]];

        // rank fn closes over the product's scoring org-set; an off-product source is devalued to -1.
        $rank = static function (array $scoringOrgs) use ($scores): callable {
            return static function (string $dim, ?string $src) use ($scores, $scoringOrgs): int {
                $base = $scores[$dim][$src] ?? 0;
                $score = in_array($src, $scoringOrgs, true) ? $base : -1;

                return -$score; // higher score => lower (better) rank
            };
        };

        $entries = [self::e('B', 'keep', 1), self::e('X', 'drop', 2)];

        // P1: X off-product -> penalised below B -> B wins.
        self::assertSame('B', Survivorship::decide('name', $entries, $rank(['A', 'B']))['winner']);

        // P2: X on-product -> B penalised -> X wins. Same claims, different context, different winner.
        self::assertSame('X', Survivorship::decide('name', $entries, $rank(['X']))['winner']);
    }
}
